File: Python/Mahasiswa.py
"""
Saya Villeneuve Andhira Suwandhi NIM 2108067 mengerjakan Latihan 2
dalam mata kuliah Desain Pemrograman Berorientasi Objek untuk keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan.
Aamiin.
"""

# Import library #
from os import system
from Sivitas import Sivitas

# Class declaration #
class Mahasiswa(Sivitas):
    # Private class
    __MHS_NUM = ""     # NIM mahasiswa
    __MHS_PRODI = ""   # Program studi
    __MHS_FAC = ""     # Fakultas

    # default constructor
    def __init__(self, n = 0, nik = "", nama = "", jenis_kelamin = "", asal_universitas = "", email_edu = "", nim = "", prodi = "", fakultas = ""):
        super().__init__(n, nik, nama, jenis_kelamin, asal_universitas, email_edu)
        self.__MHS_NUM = nim
        self.__MHS_PRODI = prodi
        self.__MHS_FAC = fakultas

    # setter
    def setMHS_ID(self, nim):
        self.__MHS_NUM = nim
    def setMHS_PRODI(self, prodi):
        self.__MHS_PRODI = prodi
    def setMHS_FAC(self, fakultas):
        self.__MHS_FAC = fakultas

    # getter
    def getMHS_ID(self):
        return self.__MHS_NUM
    def getMHS_PRODI(self):
        return self.__MHS_PRODI
    def getMHS_FAC(self):
        return self.__MHS_FAC

    # methods
    # to DISPLAY all data #
    def Show(self, m, n):
        system("cls")
        print("########## Daftar Mahasiswa ##########")
        # if there isn't data yet
        if (n == 0):
            print("!!!No data!!!")
        # if at least 1 record in data
        else:
            for i in range(n):
                print("ID                  : ", m[i].getId())
                print("NIK                 : ", m[i].getNum())
                print("NIM                 : ", m[i].getMHS_ID())
                print("NAMA                : ", m[i].getName())
                print("JENIS KELAMIN       : ", m[i].getGender())
                print("E-mail              : ", m[i].getEmailEdu())
                print("PRODI               : ", m[i].getMHS_PRODI())
                print("FAKULTAS            : ", m[i].getMHS_FAC())
                print("Asal Universitas    : ", m[i].getUniv())
                print("========================================\n")
        # new line after
        print("\n")
        # ---------------------------------------------------------------- #

    # to ADD data #
    def Insert(self):
        system("cls")
        print("### INSERT DATA ###")
        print("--- No space allowed!")
        # ID data
        self.setId(input("ID                                 : "))
        # NIK mahasiswa
        self.setNum(input("Masukkan NIK mahasiswa             : "))
        # NIM mahasiswa
        self.setMHS_ID(input("Masukkan NIM mahasiswa             : "))
        # Nama mahasiswa
        self.setName(input("Masukkan Nama mahasiswa            : "))
        # Jenis_kelamin mahasiswa
        self.setGen(input("Masukkan Jenis_kelamin mahasiswa   : "))
        # E-mail mahasiswa
        self.setEmailEdu(input("Masukkan E-mail mahasiswa          : "))
        # Program studi
        self.setMHS_PRODI(input("Masukkan Program studi mahasiswa   : "))
        # Fakultas
        self.setMHS_FAC(input("Masukkan Fakultas mahasiswa        : "))
        # Universitas mahasiswa
        self.setUniv(input("Masukkan Universitas mahasiswa     : "))
        print("\n### Data added ... ###\n\n\n")

    # ---------------------------------------------------------------- #

    # to EDIT a data by id #
    def Update(self, m, n):
        system("cls")
        # print all data
        temp = Mahasiswa()
        temp.Show(m, n)
        # init
        find = 0
        found = 0
        # find ID in data
        print("\n\n### UPDATE DATA ###")
        find = input("Masukkan ID mahasiswa yang ingin diubah :")
        # update each attribute
        for i in range(n):
            # record that has same id
            if find == m[i].getMHS_ID():
                print("\n\n### updating ... ###")
                m[i].setNum(input("Masukkan NIK baru mahasiswa             : "))
                m[i].setMHS_ID(input("Masukkan NIM baru mahasiswa             : "))
                m[i].setName(input("Masukkan Nama baru mahasiswa            : "))
                m[i].setGen(input("Masukkan Jenis_kelamin baru mahasiswa   : "))
                m[i].setEmailEdu(input("Masukkan E-mail baru mahasiswa          : "))
                m[i].setMHS_PRODI(input("Masukkan Program studi baru mahasiswa   : "))
                m[i].setMHS_FAC(input("Masukkan Fakultas baru mahasiswa        : "))
                m[i].setUniv(input("Masukkan Universitas baru mahasiswa     : "))
                # sign that the current record was updated
                print("\n\n### Data updated ###\n\n")
                # it means that the record has found
                found = 1
                return
            
        # if the ID is not found in data
        if found == 0:
            print("### ID is not found ###")

    # ---------------------------------------------------------------- #

    # to DELETE a data by id #
    def Delete(self, m, n):
        system("cls")
        # print all data
        temp = Mahasiswa()
        temp.Show(m, n)
        # init
        find = 0
        found = 0
        # find ID in data
        print("\n\n### DELETE DATA ###")
        find = input("Masukkan ID mahasiswa yang ingin dihapus :")
        for i in range(n):
            # record that has same id
            if find == m[i].getMHS_ID():
                # shift record to maks
                for j in range(i, n-1):
                    m[j] = m[j+1]
                # sign that the current record was deleted
                print("\n\n### Data successfully deleted ###\n\n")
                # it means that the record has found
                found = 1
                return
        
        # if the ID is not found in data
        if found  == 0:
            print("\n\n### ID is not found ###\n\n")

File: C++/Mahasiswa.cpp
/*
Saya Villeneuve Andhira Suwandhi NIM 2108067 mengerjakan Latihan 2
dalam mata kuliah Desain Pemrograman Berorientasi Objek untuk keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan.
Aamiin.
*/

/* Import library */
#include <iostream>
#include <string>
#include "Sivitas.cpp"

/* Using standard namespace */
using namespace std;

/* Class declaration */
class Mahasiswa : public Sivitas {
    private:
        string MHS_NUM;     // NIM mahasiswa
        string MHS_PRODI;   // Program studi
        string MHS_FAC;     // Fakultas

    public:
        // default constructor
        Mahasiswa() {
            this->MHS_NUM = "";
            this->MHS_PRODI = "";
            this->MHS_FAC = "";
        }

        //constructor
        Mahasiswa(int n, string nik, string nama, string jenis_kelamin, string asal_universitas, string email_edu, string nim, string prodi, string fakultas) {
            Sivitas(n, nik, nama, jenis_kelamin, asal_universitas, email_edu);
            this->MHS_NUM = nim;
            this->MHS_PRODI = prodi;
            this->MHS_FAC = fakultas;
        }

        // setter
        void setMHS_ID(string nim) {
            this->MHS_NUM = nim;
        }
        void setMHS_PRODI(string prodi) {
            this->MHS_PRODI = prodi;
        }
        void setMHS_FAC(string fakultas) {
            this->MHS_FAC = fakultas;
        }

        // getter
        string getMHS_ID() {
            return this->MHS_NUM;
        }
        string getMHS_PRODI(){
            return this->MHS_PRODI;
        }
        string getMHS_FAC(){
            return this->MHS_FAC;
        }

        // methods
        /* to DISPLAY all data */
        void Show(Mahasiswa m[], int n) {
            system("cls");
            cout << "########## Daftar Mahasiswa ##########" << endl;
            // if there isn't data yet
            if (n == 0) {
                cout << "!!!No data!!!" << endl;
            }
            // if at least 1 record in data
            else {
                for (int i = 0; i < n; i++) {
                    cout << "ID                  : " << m[i].getId() << endl;
                    cout << "NIK                 : " << m[i].getNum() << endl;
                    cout << "NIM                 : " << m[i].MHS_NUM << endl;
                    cout << "NAMA                : " << m[i].getName() << endl;
                    cout << "JENIS KELAMIN       : " << m[i].getGender() << endl;
                    cout << "E-mail              : " << m[i].getEmailEdu() << endl;
                    cout << "PRODI               : " << m[i].MHS_PRODI << endl;
                    cout << "FAKULTAS            : " << m[i].MHS_FAC << endl;
                    cout << "Asal Universitas    : " << m[i].getUniv() << endl;
                    cout << "========================================" << endl;
                }
            }
            // new line after
            cout << "\n" << endl;
        }

        /* to ADD data */
        void Insert(Mahasiswa m[], int n) {
            system("cls");
            // Instantiate
            int intInput;
            char strInput[99];
            cout << "### INSERT DATA ###" << endl;
            cout << "--- No space allowed!" << endl;
            // ID data
            cout << "ID                                 : ";
            cin >> intInput;
            m[n].setId(intInput);
            // NIK mahasiswa
            cout << "Masukkan NIK mahasiswa             : ";
            cin >> strInput;
            m[n].setNum(strInput);
            // NIM mahasiswa
            cout << "Masukkan NIM mahasiswa             : ";
            cin >> strInput;
            m[n].setMHS_ID(strInput);
            // Nama mahasiswa
            cout << "Masukkan Nama mahasiswa            : ";
            cin >> strInput;
            m[n].setName(strInput);
            // Jenis_kelamin mahasiswa
            cout << "Masukkan Jenis_kelamin mahasiswa   : ";
            cin >> strInput;
            m[n].setGen(strInput);
            // E-mail mahasiswa
            cout << "Masukkan E-mail mahasiswa          : ";
            cin >> strInput;
            m[n].setEmailEdu(strInput);
            // Program studi mahasiswa
            cout << "Masukkan Program studi mahasiswa   : ";
            cin >> strInput;
            m[n].setMHS_PRODI(strInput);
            // Fakultas mahasiswa
            cout << "Masukkan Fakultas mahasiswa        : ";
            cin >> strInput;
            m[n].setMHS_FAC(strInput);
            // Universitas mahasiswa
            cout << "Masukkan Universitas mahasiswa     : ";
            cin >> strInput;
            m[n].setUniv(strInput);
            // notification
            cout << "\n### Data added ... ###\n" << endl;
            cout << "\n\n";
        }

        /* to EDIT a data by id */
        void Update(Mahasiswa m[], int n) {
            system("cls");
            // print all data
            Show(m, n);
            // instantiate
            int find, found;
            int intInput;
            char strInput[99];
            // find ID in data
            cout << "\n\n### UPDATE DATA ###" << endl;
            cout << "Masukkan ID mahasiswa yang ingin diubah :";
            cin >> find;
            // init
            found = 0;
            // update each attribute
            for (int i = 0; i < n; i++) {
                // record that has same id
                if (find == m[i].getId()) {
                    cout << "\n\n### updating ... ###" << endl;
                    // NIK mahasiswa
                    cout << "Masukkan NIK baru mahasiswa            : ";
                    cin >> strInput;
                    m[i].setNum(strInput);
                    // NIM mahasiswa
                    cout << "Masukkan NIM baru mahasiswa            : ";
                    cin >> strInput;
                    m[i].setMHS_ID(strInput);
                    // Nama mahasiswa
                    cout << "Masukkan Nama baru mahasiswa           : ";
                    cin >> strInput;
                    m[i].setName(strInput);
                    // Jenis_kelamin mahasiswa
                    cout << "Masukkan Jenis_kelamin baru mahasiswa  : ";
                    cin >> strInput;
                    m[i].setGen(strInput);
                    // E-mail mahasiswa
                    cout << "Masukkan E-mail baru mahasiswa         : ";
                    cin >> strInput;
                    m[i].setEmailEdu(strInput);
                    // Program studi mahasiswa
                    cout << "Masukkan Program studi baru mahasiswa  : ";
                    cin >> strInput;
                    m[i].setMHS_PRODI(strInput);
                    // Fakultas mahasiswa
                    cout << "Masukkan Fakultas baru mahasiswa       : ";
                    cin >> strInput;
                    m[i].setMHS_FAC(strInput);
                    // Universitas mahasiswa
                    cout << "Masukkan Universitas baru mahasiswa    : ";
                    cin >> strInput;
                    m[i].setUniv(strInput);
                    // sign that the current record was updated
                    cout << "\n\n### Data updated ###" << endl;
                    // new line after
                    cout << "\n\n";
                    // it means that the record has found
                    found = 1;
                    break;
                }
            }
            // if the ID is not found in data
            if (found == 0) {
                cout << "### ID is nout found ###" << endl;
            }
        }

        /* to DELETE a data by id */
        void Delete(Mahasiswa m[], int &n) {
            system("cls");
            // print all data
            Show(m, n);
            // instantiate
            int find, found;
            // find ID in data
            cout << "\n\n### DELETE DATA ###" << endl;
            cout << "Masukkan ID mahasiswa yang ingin dihapus :";
            cin >> find;
            // init
            found = 0;
            for (int i = 0; i < n; i++) {
                // record that has same id
                if (find == m[i].getId()) {
                    // shift record to maks
                    for (int j = i; j < n; j++) {
                        m[j] = m[j+1];
                    }
                    // delete record
                    n--;
                    // sign that the current record was deleted
                    cout << "\n\n### Data successfully deleted ###" << endl;
                    // new line after
                    cout << "\n\n";
                    // it means that the record has found
                    found = 1;
                    break;
                }
            }
            // if the ID is not found in data
            if (found  == 0) {
                cout << "\n\n### ID is nout found ###" << endl;
                cout << "\n\n";
            }

        }
    
        // destructor
        ~Mahasiswa() {
        }
};